src/mime.cr

```diff
@@ -0,0 +1,169 @@
+# MIME.extensions("text/crystal") # => [".cr"]
+# ```
+module MIME
+  class Error < Exception
+  end
+
+  @@initialized = false
+  @@types = {} of String => String
+  @@types_lower = {} of String => String
+  @@extensions = {} of String => Array(String)
+
+  # A limited set of default MIME types.
+  DEFAULT_TYPES = {
+    ".css"  => "text/css; charset=utf-8",
+    ".gif"  => "image/gif",
+    ".htm"  => "text/html; charset=utf-8",
+    ".html" => "text/html; charset=utf-8",
+    ".jpg"  => "image/jpeg",
+    ".jpeg" => "image/jpeg",
+    ".js"   => "application/x-javascript",
+    ".pdf"  => "application/pdf",
+    ".png"  => "image/png",
+    ".svg"  => "image/svg+xml",
+    ".txt"  => "text/plain; charset=utf-8",
+    ".xml"  => "text/xml; charset=utf-8",
+  }
+
+  private def self.initialize_types
+    return if @@initialized
+    @@initialized = true
+
+    DEFAULT_TYPES.each do |ext, type|
+      register ext, type
+    end
+
+    Crystal::System::MIME.load
+  end
+
+  # Looks up the MIME type associated with *extension*.
+  #
+  # A case sensitive search is tried first, if this yields not result, it it
+  # matched case-insensitive. Returns *default* if *extension* is not registered.
+  def self.fetch(extension : String, default) : String
+    fetch(extension) { default }
+  end
+
+  # Looks up the MIME type associated with *extension*.
+  #
+  # A case sensitive search is tried first, if this yields not result, it it
+  # matched case-insensitive. Raises `KeyError` if *extension* is not registered.
+  def self.fetch(extension : String) : String
+    fetch(extension) { raise KeyError.new("Missing MIME type for extension #{extension.inspect}") }
+  end
+
+  # Looks up the MIME type associated with *extension*.
+  #
+  # A case sensitive search is tried first, if this yields not result, it it
+  # matched case-insensitive. Returns `nil` if *extension* is not registered.
+  def self.fetch?(extension : String) : String?
+    fetch(extension) { nil }
+  end
+
+  # Looks up the MIME type associated with *extension*.
+  #
+  # A case sensitive search is tried first, if this yields not result, it it
+  # matched case-insensitive. Runs the fiven block if *extension* is not registered.
+  def self.fetch(extension : String, &block)
+    initialize_types
+
+    @@types.fetch(extension) { @@types_lower.fetch(extension.downcase) { yield extension } }
+  end
+
+  # Register *type* for *extension*.
+  #
+  # *extension* must start with a dot (`.`) and must not contain any null bytes.
+  def self.register(extension : String, type : String) : Nil
+    raise ArgumentError.new("extension does not start with a dot: #{extension.inspect}") unless extension.starts_with?('.')
+    extension = extension.check_no_null_byte.downcase
+
+    initialize_types
+
+    mediatype = parse_media_type(type) || raise Error.new "invalid media type: #{type}"
+
+    @@types[extension] = type
+    @@types_lower[extension.downcase] = type
```

Comment:
But you *always* downcase on line 99.

Reply:
Yeah, it shouldn't be downcased before.

Reply:
Why bother keeping two hashes, and not just `downcase`-d one?

Reply:
@Sija Resolution should be case-sensitive by default and only fall back to case-insensitive if there is no match.